common/net/credentials/alts/handshaker/handshaker.go

```diff
@@ -0,0 +1,373 @@
+	if _, err := s.conn.Write(d); err != nil {
+		return fmt.Errorf("write handshake record: %w", err)
+	}
+	return nil
+}
+
+func (s *altsHandshaker) doClientHandshake(ctx context.Context, auth alts.Authentication, signInfo *alts.SignInfo) error {
+	curve := ed448.NewCurve()
+	// generate the private and public key for client
+	priv, pub, ok := curve.GenerateKeys()
+	if !ok {
+		return errors.New("failed to generate keys")
+	}
+
+	signature, err := auth.Sign(ctx, pub[:], signInfo.PastelID, signInfo.PassPhrase)
+	if err != nil {
+		return fmt.Errorf("failed to generate signature: %w", err)
+	}
+
+	request := kexExchangeRequest{
+		PubKey:    pub[:],
+		Signature: signature,
+		PastelID:  signInfo.PastelID,
+	}
+	// encode and write the client's handshake record
+	if err := s.writeHandshake(&request); err != nil {
+		return fmt.Errorf("write handshake: %w", err)
+	}
+
+	// read and decode the client's handshake record
+	var response kexExchangeResponse
+	if err := s.readHandshake(&response); err != nil {
+		return fmt.Errorf("read handshake: %w", err)
+	}
+
+	if ok, err := auth.Verify(ctx, response.PubKey, string(response.Signature), response.PastelID); err != nil || !ok {
+		return fmt.Errorf("failed to verify server public key: %w", err)
+	}
+
+	var serverPubKey [ED448PubKeySize]byte
+	copy(serverPubKey[:], response.PubKey)
+	// compute the secret key for connection
+	secret := curve.ComputeSecret(priv, serverPubKey)
+
+	subkeyLen, ok := recordKeyLen[s.protocol]
+	if !ok {
+		return fmt.Errorf("unknown resulted record protocol: %v", s.protocol)
+	}
+
+	kdfValue, exit := cryptokdf.CryptoKdfDeriveFromKey(subkeyLen, kdfSubKeyID, kdfContext,
+		secret[:cryptokdf.CryptoKdfKeybytes()])
+	if exit != 0 {
+		return fmt.Errorf("failed to compute kdf, exit=%d", exit)
+	}
+	// update the record key
+	s.key = kdfValue
+
+	return nil
+}
+
+// ClientHandshake starts and completes a client ALTS handshaking. Once
+// done, ClientHandshake returns a secure connection.
+func (s *altsHandshaker) ClientHandshake(ctx context.Context, auth alts.Authentication, signInfo *alts.SignInfo) (net.Conn, credentials.AuthInfo, error) {
+	if !acquire() {
+		return nil, nil, errDropped
+	}
+	defer release()
+
+	if s.side != alts.ClientSide {
+		return nil, nil, errors.New("only handshakers created using NewClientHandshaker can perform a client handshaker")
+	}
+
+	// do the client handshake
+	if err := s.doClientHandshake(ctx, auth, signInfo); err != nil {
+		return nil, nil, fmt.Errorf("do client handshake: %w", err)
+	}
+	log.WithContext(ctx).Debugf("client handshake is complete")
+
+	// new a secure connection
+	sc, err := conn.NewConn(s.side, s.conn, s.protocol, s.key)
+	if err != nil {
+		return nil, nil, fmt.Errorf("new secure conn: %w", err)
+	}
+
+	return sc, authinfo.New(), nil
+}
+
+func (s *altsHandshaker) doServerHandshake(ctx context.Context, auth alts.Authentication, signInfo *alts.SignInfo) error {
+	// read and decode the client's handshake record
```

Comment:
this handshake is not robust enough.
If they can have valid public key/private key+ signature of one pastelId (without passphrase) (maybe via one valid transaction) - they can fake it.
Should have a challenge token included in exchange process.
Like:
1. client - send a challengeA (randomily) to sever
2. sever - send a challenceB(randomly) to client
3. client : generate pub/secret, sign concat(its pub, challenceB), then send (its pub,signature)
4. sever: verify signature, then also sign concat(its pub, challenceA), then send (its pub, signature)
5. client: verify signature

Reply:
ahh. i got your idea.